### DISCUSSION TITLE
tutorials/scripting/c_sharp/c_sharp_basics

### DISCUSSION THREAD

**Discussion by giscus** [godotengine/godot-docs-user-notes#38] [User-contributed notes]
# tutorials/scripting/c_sharp/c_sharp_basics

Introduction: This page provides a brief introduction to C#, both what it is and how to use it in Godot. Afterwards, you may want to look at how to use specific features, read about the differences...

https://docs.godotengine.org/en/stable/tutorials/scripting/c_sharp/c_sharp_basics.html

**Comment by EricPendergast** [comment]
The example `.csproj` for the "Using NuGet packages in Godot" is somewhat misleading since newtonsoft is currently broken in the editor per https://github.com/godotengine/godot/issues/78513

**Comment by ondesic** [comment]
Very important. When converting from 3.x to 4.x, Newtonsoft would not work. It couldn't be found. I noticed that the old .mono folder was in the main project folder. I closed all editors and code IDEs, renamed .mono  to .monoOLD and restarted it. had to reimport Newtonsoft in the Rider|NuGut manager and then it finally recognized it!

**Comment by dreamReason** [comment]
In order to see the DotNet tree to configure the editor in 4.4.1, you must turn on 'advanced settings'

**Comment by shyamjayakannan** [comment]
I found using `using System.Text.Json;` to be a good replacement for NewtonSoft

**Comment by ondesic** [comment]
I stopped using System.Text.Json because though it works on Windows, it
crashes on Android. Has this been fixed?

**Reply by shyamjayakannan** [reply]
how are you exporting to android in the first place? as far as i remember it's still experimental for c# projects as of godot 4.4.

**Comment by ondesic** [comment]
I have had no problem exporting for Android

**Comment by ondesic** [comment]
I have the .net version.

**Comment by sappqa** [comment]
One thing worth noting if you're having issues finding the 'Dotnet' option in the editor settings, make sure 'Advanced Options' is enabled, then the 'Dotnet' option should appear at the bottom of the left scroll menu.

**Comment by Nidvogr** [comment]
I found it unclear whether the newest version of godot (4.5), requires .net or mono, the download page itself says mono, but the mono runtime is not the same as .net 9 which should be the consolidated cross platform version. This page itself talks about godot being based on .net 8. So does godot require mono, or use mono, or use .net 8?

**Reply by Calinou** [reply]
You need .NET 9 with Godot 4.5 to be specific. The Mono runtime isn't used in Godot 4 anymore (it was used in Godot 3).

Godot's C# integration is still called `mono` for legacy reasons, but it's fully based on .NET as of now.

Pull request opened:

- https://github.com/godotengine/godot-docs/pull/11348

**Reply by raulsntos** [reply]
As @Calinou said, Godot's C# integration is fully based on modern .NET hosting since 4.0, we no longer use the Mono embedding APIs. However, to be clear, modern .NET includes the Mono runtime which was merged into .NET in 5.0.

Godot itself uses the Mono runtime for Android exports, since there is no CoreCLR support for Android yet. For iOS we use NativeAOT. And for all other supported platforms (Windows, Linux, and macOS) we use CoreCLR by default but other runtimes (Mono and NativeAOT) are also supported. But you don't have to worry about which runtimes are supported, just install the latest .NET SDK and you're good to go.

As for which specific version of .NET is required, the minimum should be .NET 8 since that's what Godot's packages target. However, on Android we require .NET 9 to support the [latest Google Play requirements](https://android-developers.googleblog.com/2025/05/prepare-play-apps-for-devices-with-16kb-page-size.html). But your project can always target a newer version if you want.

Regardless of which version your project targets, it's still always recommended to use the latest .NET SDK to build your projects. So, as of now, that would be .NET 9, but feel free to update to .NET 10 this November.

**Comment by zyphermann** [comment]
On my macbook (time of writing is 17/11/2025) I had to modify the tasks.json like this: 
```json
{
  "version": "2.0.0",
  "tasks": [
    {
      "label": "build",
      "command": "dotnet",
      "type": "process",
      "args": ["build"],
      "problemMatcher": "$msCompile",
      "options": {
        "env": {
          "PATH": "/usr/local/share/dotnet:${env:PATH}"
        }
      }
    }
  ]
} 
```

Without the options.env section it would not work. 

Launch JSON looks like so. Could not find a solution for "${env:GODOT4}" seems this does not work on MacOS.


```json
{
  "version": "0.2.0",

  "configurations": [
    {
      "name": "Play",
      "type": "coreclr",
      "request": "launch",
      "preLaunchTask": "build",
      "program": "/Applications/Godot_mono_4.5.app/Contents/MacOS/Godot",
      "args": [],
      "cwd": "${workspaceFolder}",
      "stopAtEntry": false
    }
  ]
}
```

Another step I had to perform was

```
ln -s /usr/local/share/dotnet/dotnet /usr/local/bin/
```

Now I can fully debug from VSCode for the first time. Took me hours to figure out, so I would like to share.
Using Godot 4.5 c# on MacOS Ventura 13.2.1
